File: Python/paradigmas/1029fibonacciquantaschamadas.py
from collections import OrderedDict
fib = {
    1: OrderedDict({
        'num': 0,
        'call': 1
    }),
    2: OrderedDict({
        'num': 2,
        'call': 1
    }),
    3: OrderedDict({
        'num': 4,
        'call': 2
    }),
    4: OrderedDict({
        'num': 8,
        'call': 3
    }),
}
for i in range(5, 40):
    fib[i] = OrderedDict({
        'num': fib[i - 1]['num'] + fib[i - 2]['num'] + 2,
        'call': fib[i - 1]['call'] + fib[i - 2]['call']
    })

j = int(input())
for i in range(j):
    k = int(input())
    res = fib[k]
    chamada = res['num']
    res = res['call']
    print('fib(' + str(k) + ") = " + str(chamada) + ' calls = ' + str(res))

File: Haskell/paradigmas/1029fibonacciquantaschamadas.hs
import           Text.Printf


fib :: [(Int, Int)]
fib = [(0, 1), (2, 1), (4, 2), (8, 3)]


geraFib :: Int -> [(Int, Int)] -> [(Int, Int)]
geraFib i l = if i <= 39
    then geraFib
        (i + 1)
        (  l
        ++ [ ( fst (l !! (i - 2)) + fst (l !! (i - 3)) + 2
             , snd (l !! (i - 2)) + snd (l !! (i - 3))
             )
           ]
        )
    else l

listaFib = geraFib 5 fib

executa :: Int -> IO ()
executa 0   = return ()
executa num = do
    n <- readLn :: IO Int
    let nl = listaFib !! (n - 1)
    uncurry (printf "fib(%d) = %d calls = %d\n" n) nl
    executa (num - 1)

main :: IO ()
main = do
    n <- readLn :: IO Int
    executa n
    return ()
